src/textual/widgets/_progress_bar.py

```diff
@@ -0,0 +1,432 @@
+
+    def render(self) -> RenderResult:
+        """Render the bar with the correct portion filled."""
+        if self._completion_percentage is None:
+            return self.render_indeterminate()
+        else:
+            bar_style = (
+                self.get_component_rich_style("bar--bar")
+                if self._completion_percentage < 1
+                else self.get_component_rich_style("bar--complete")
+            )
+            return RenderableBar(
+                highlight_range=(0, self.size.width * self._completion_percentage),
+                highlight_style=Style.from_color(bar_style.color),
+                background_style=Style.from_color(bar_style.bgcolor),
+            )
+
+    def render_indeterminate(self) -> RenderResult:
+        """Render a frame of the indeterminate progress bar animation."""
+        width = self.size.width
+        highlighted_bar_width = 0.25 * width
+        # Width used to enable the visual effect of the bar going into the corners.
+        total_imaginary_width = width + highlighted_bar_width
+
+        speed = 30  # Cells per second.
+        # Compute the position of the bar.
+        start = (speed * monotonic()) % (2 * total_imaginary_width)
+        if start > total_imaginary_width:
+            # If the bar is to the right of its width, wrap it back from right to left.
+            start = 2 * total_imaginary_width - start  # = (tiw - (start - tiw))
+        start -= highlighted_bar_width
+        end = start + highlighted_bar_width
+
+        bar_style = self.get_component_rich_style("bar--indeterminate")
+        return RenderableBar(
+            highlight_range=(max(0, start), min(end, width)),
+            highlight_style=Style.from_color(bar_style.color),
+            background_style=Style.from_color(bar_style.bgcolor),
+        )
+
+
+class PercentageStatus(Label):
+    """A label to display the percentage status of the progress bar."""
+
+    _completion_percentage: reactive[float | None] = reactive[Optional[float]](None)
+    """The percentage of progress that has been completed or `None` if indeterminate."""
+
+    DEFAULT_CSS = """
+    PercentageStatus {
+        width: 5;
+        content-align-horizontal: right;
+    }
+    """
+
+    def watch__completion_percentage(self, percentage: float | None) -> None:
+        if percentage is None:
+            self.renderable = "--%"
+        else:
+            self.renderable = f"{int(100 * percentage)}%"
+
+
+class ETAStatus(Label):
+    """A label to display the estimated time until completion of the progress bar."""
+
+    _completion_percentage: reactive[float | None] = reactive[Optional[float]](None)
+    """The percentage of progress that has been completed or `None` if indeterminate."""
+
+    DEFAULT_CSS = """
+    ETAStatus {
+        width: 9;
+        content-align-horizontal: right;
+    }
+    """
+
+    _refresh_timer: Timer | None = None
```

Comment:
Not keen on setting initial values as classvars. Initial values should be the domain of `__init__`.